File: Testing/Core/vtkEmscriptenTestUtilities.cxx
// SPDX-FileCopyrightText: Copyright (c) Ken Martin, Will Schroeder, Bill Lorensen
// SPDX-License-Identifier: BSD-3-Clause
#ifdef __EMSCRIPTEN__

#include "vtkEmscriptenTestUtilities.h"

#include <cstdlib>                // for free
#include <emscripten.h>           // for EM_ASM macro
#include <iostream>               // for cerr, cout
#include <vtksys/SystemTools.hxx> // for filesystem utilities

VTK_ABI_NAMESPACE_BEGIN

/**
 * Given a path to a file "a/b/c/d.ext" on the server hosting webassembly runtime/page,
 * this method preloads it into the sandbox at the given `sandboxedFilePath`
 */
void vtkEmscriptenTestUtilities::PreloadDataFile(
  const std::string& hostFilePath, const std::string& sandboxedFilePath)
{
  FILE* f = nullptr;
  // This function preloads the file and it's size in bytes into wasm memory from javascript.
  // It returns a pointer to a PreloadData.
  auto* payload =
    reinterpret_cast<PreloadDescriptor*>(vtkPreloadDataFileIntoMemory(hostFilePath.c_str()));
  if (payload != nullptr)
  {
    const auto dirName = vtksys::SystemTools::GetFilenamePath(sandboxedFilePath);
    if (!dirName.empty() && dirName != "/")
    {
      const auto status = vtksys::SystemTools::MakeDirectory(dirName);
      if (!status.IsSuccess())
      {
        std::cerr << "Failed to create parent directory for " << sandboxedFilePath << '\n';
        free(payload->buffer);
        free(payload);
        return;
      }
      else
      {
        std::cout << "Created directory for " << sandboxedFilePath << '\n';
      }
    }
    f = fopen(sandboxedFilePath.c_str(), "wb+");
    if (f != nullptr)
    {
      auto written = fwrite(payload->buffer, sizeof(payload->buffer[0]), payload->size, f);
      fclose(f);
      if (written != payload->size)
      {
        std::cerr << "Failed to preload " << payload->size << " bytes from " << hostFilePath
                  << " into " << sandboxedFilePath << "\n";
      }
      else
      {
        std::cout << "Preloaded " << payload->size << " bytes from " << hostFilePath << " into "
                  << sandboxedFilePath << "\n";
      }
    }
    else
    {
      std::cerr << "Failed to create sandboxed " << sandboxedFilePath << " for " << hostFilePath
                << '\n';
    }
    free(payload->buffer);
    free(payload);
  }
}

/**
 * Given a path to a file "a/b/c/d.ext" on the server hosting webassembly runtime/page,
 * this method preloads it into the sandbox at "/d.ext" and returns "/d.ext"
 */
std::string vtkEmscriptenTestUtilities::PreloadDataFile(const std::string& hostFilePath)
{
  std::string sandboxedFilePath = vtksys::SystemTools::GetFilenameName(hostFilePath);
  vtkEmscriptenTestUtilities::PreloadDataFile(hostFilePath, sandboxedFilePath);
  return sandboxedFilePath;
}

/**
 * This method helps to write files outside of the webassembly sandbox.
 * Writes `n` number of bytes from `data` into a file called `hostFilePath` on the server's file
 * system. The server is a system which hosts the webassembly runtime/page.
 */
void vtkEmscriptenTestUtilities::DumpFile(
  const std::string& hostFilePath, const void* data, size_t n)
{
  const auto* bytes = reinterpret_cast<const uint8_t*>(data);
  vtkDumpFile(hostFilePath.c_str(), bytes, n);
}
VTK_ABI_NAMESPACE_END
#endif // __EMSCRIPTEN__

File: Testing/Core/vtkEmscriptenTestUtilities.js
// SPDX-FileCopyrightText: Copyright (c) Ken Martin, Will Schroeder, Bill Lorensen
// SPDX-License-Identifier: BSD-3-Clause
/**
 * Define utility functions to preload and dump a file.
 * This file gets injected as a `--pre-js` file.
 */

var vtkEmscriptenTestUtilities = {
  /**
   * Fetches a file `hostPath` from server and copies webassembly memory.
   * Here is how you can call this from C/C++ code:
   * ```cpp
   * #include "vtkEmscriptenTestUtilities.h" // for PreloadDescriptor and vtkPreloadDataFileIntoMemory
   * vtkEmscriptenTestUtilities::PreloadDescriptor* payload = nullptr;
   * payload = reinterpret_cast<vtkEmscriptenTestUtilities::PreloadDescriptor*>(vtkPreloadDataFileIntoMemory("/path/to/server/directory/file.png"));
   * ```
   * @param {*} hostPath A valid path in the server"s filesystem.
   */
  vtkPreloadDataFileIntoMemory__deps: ['malloc'],
  vtkPreloadDataFileIntoMemory__sig: "pp",
  vtkPreloadDataFileIntoMemory: (hostPathFile) => {
    hostPathFile = UTF8ToString(hostPathFile);
    const req = new XMLHttpRequest;
    req.overrideMimeType('text/plain; charset=x-user-defined');
    req.open("GET", `preload?file=${hostPathFile}`, false);
    req.send(null);
    if (!(req.status >= 200 && req.status < 300)) {
      return 0;
    } else if (req.response !== undefined) {
      let data = Uint8Array.from(req.response, c => c.charCodeAt(0));
      const fileContentPtr = _malloc(data.length);
      HEAPU8.set(data, fileContentPtr);
      const payloadPtr = _malloc(4 + {{{ POINTER_SIZE }}});
      {{{ makeSetValue('payloadPtr', 0, 'fileContentPtr', '*') }}}
      {{{ makeSetValue('payloadPtr', `${POINTER_SIZE}`, 'data.length', '*') }}}
      return payloadPtr;
    }
  },

  /**
   * Helps VTK C++ unit tests write a file in the server"s file system.
   * Here is how you can call this from C/C++ code:
   * ```cpp
   * #include "vtkEmscriptenTestUtilities.h" // for vtkDumpFile
   * vtkDumpFile("/path/to/server/directory/file.png", bytes, number_of_bytes);
   * ```
   * @param {*} hostPath A valid path in the server"s filesystem.
   * @param {*} data A Uint8Array representing the contents of the file which will be dumped out in the server"s filesystem.
   */
  vtkDumpFile__sig: "vppp",
  vtkDumpFile: (hostPathFile, data, nbytes) => {
    hostPathFile = UTF8ToString(hostPathFile);
    const req = new XMLHttpRequest;
    req.open("POST", `dump?file=${hostPathFile}`);
    const byteArray = HEAPU8.subarray(data, data + nbytes);
    req.send(new Uint8Array(byteArray));
  },

  /**
   * Send the exit code to the server.
   */
  vtkPostExitCode__sig: "vi",
  vtkPostExitCode: (code) => {
    // this function is defined in Testing/WebAssembly/templates/index.html
    finalize(code)
  },
};

mergeInto(LibraryManager.library, vtkEmscriptenTestUtilities);
